### PR TITLE
Distributed sampling parity between Lite and PyTorch

### LINKED ISSUE #14636: Introduce a mechanism to set the epoch on the sampler in LightningLite
## 🚀 Feature

We need a mechanism to set the epoch on the distributed sampler via `.set_epoch()`. 

### Motivation

To correctly handle shuffling with the DistributedSampler in DDP, the PyTorch user would normally call

```py
sampler.set_epoch(epoch)
```

in their training loop.  PL handles this automatically for the user, but in Lite, the training loop is owned by the user and hence should be signaled in a different way. 

### Pitch

Provide a strategy-agnostic API to set the epoch. The idea here is that you don't need to change the code or add boilerplate conditional logic to handle the sampler when you switch from DDP to single device or vice versa. 

Before

```py
....
train_dataloader, val_dataloader = self.setup_dataloaders(train_dataloader, val_dataloader)

for epoch in range(num_epochs):

    # Beginning of a new epoch

    # Boilerplate code
    if isinstance(train_dataloader.sampler, DistributedSampler):
        train_dataloader.sampler.set_epoch(epoch)


    for idx, data in enumerate(train_dataloader):
        ...

```

Now:

```py
....
train_dataloader, val_dataloader = self.setup_dataloaders(train_dataloader, val_dataloader)

for epoch in range(num_epochs):

    # Beginning of a new epoch

    # This is a no-op if not using distributed sampler (DDP)
    self.set_epoch(epoch, train_dataloader)

    for idx, data in enumerate(train_dataloader):
        ...
```

### Alternatives

Don't introduce this. It is left to the user to handle this boilerplate code. 

### Additional context

[PyTorch docs for set_epoch
](https://pytorch.org/docs/stable/data.html#torch.utils.data.distributed.DistributedSampler)

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Restores parity between pure PyTorch dataloader and Lite dataloader when using DistributedSampler (DDP). 
This PR avoids the DistributedSamplerWrapper for the two standard samplers, RandomSampler and SequentialSampler. The DistributedSamplerWrapper should only be used for custom samplers. A note was added to the documentation of this wrapper.

We also ensure that `.set_epoch()` gets called automatically for the user. 

Fixes #14636

### Does your PR introduce any breaking changes? If yes, please list them.

The order of samples when shuffling is enabled. But this should be considered a bugfix.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: src/lightning_lite/wrappers.py [hunk 1]
```diff
@@ -160,6 +161,13 @@ def __len__(self) -> int:
         return len(self._dataloader)
 
     def __iter__(self) -> Union[Iterator[Any], Generator[Any, None, None]]:
+        if hasattr(self._dataloader.sampler, "set_epoch"):
+            # Without setting the epoch, the distributed sampler would return the same indices every time, even when
+            # shuffling is enabled. In PyTorch, the user would normally have to call `.set_epoch()` on the sampler.
+            # In Lite, we take care of this boilerplate code.
+            self._dataloader.sampler.set_epoch(self._epoch)
```
Comment (tchaton) [inline src/lightning_lite/wrappers.py:168]: Neat !

Comment (tchaton) [reply src/lightning_lite/wrappers.py:168]: Should we add a log info there ? So users don't set_epoch too ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning_lite/wrappers.py (hunk 1 above)
Comment (justusschock) [reply src/lightning_lite/wrappers.py:168]: if we do, it should at max be a log debug :) Otherwise this will become very verbose

Comment (awaelchli) [reply src/lightning_lite/wrappers.py:168]: I don't think there is a need to inform. We also don't do it in Lightning Trainer.
It's fine if the user also calls `set_epoch` manually.

File: tests/tests_lite/test_lite.py [hunk 2]
```diff
@@ -405,7 +405,46 @@ def test_setup_dataloaders_distributed_sampler_shuffle():
...
+
+@pytest.mark.parametrize("shuffle", [True, False])
+@pytest.mark.parametrize("batch_size", [1, 2, 3])
+def test_setup_dataloaders_distributed_sampler_parity(shuffle, batch_size):
+    """Test that the distributed sampler setup in Lite leads to the same sequence of data as in raw PyTorch."""
+    torch.manual_seed(1)
+    lite = LightningLite(accelerator="cpu", strategy="ddp", devices=2)
+    # no lite.launch(): pretend we are on rank 0 now
+
+    dataset = torch.rand(10, 1)
```
Comment (carmocca) [inline tests/tests_lite/test_lite.py:419]: nit: This test could be simpler if you used `torch.arange` instead of a random tensor

Comment (awaelchli) [reply tests/tests_lite/test_lite.py]: Sure, it does not matter, what the tensor is. I couldn't find a way to make the test simpler

File: src/lightning_lite/wrappers.py [hunk 3]
```diff
@@ -151,6 +151,7 @@ def __init__(self, dataloader: DataLoader, device: Optional[torch.device] = None
         self.__dict__.update(dataloader.__dict__)
         self._dataloader = dataloader
         self._device = device
+        self._epoch = 0
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning_lite/wrappers.py]: Also in the distributed sampler from torch, the definition of epoch there must not be what the users defines in their loop. It's has to be purely defined by how many times the iterator gets recreated.
